jgrapht-core/src/main/java/org/jgrapht/alg/interval/KorteMoehringIntervalGraphRecognizer.java

```diff
@@ -97,163 +145,196 @@ private void testIntervalGraph()
+        for (int i = 0; i < vertexOrder.size(); i++) {
+            vertexIndexMap.put(vertexOrder.get(i), i);
         }
-        return vertexInOrder;
+        return vertexIndexMap;
     }
-    
-    
+
     /**
      * Changed the MPQ Tree if u has no predecessors.
      * Adds a new leaf node with the bag of this vertex to the root.
      * 
      * @param u the vertex to be added to the MPQ Tree
      */
-    private void addEmptyPredecessors(V u)
-    {
-        MPQNodeSetElement bag = new MPQNodeSetElement(u);
-        MPQNode leaf = new PNode(null,bag);
-        treeRoot.add(leaf);
+    private void addEmptyPredecessors(V u) {
+        HashSet<V> elements = new HashSet<>();  // to be implemented by the doubly linked circular list
+        elements.add(u);
+        MPQTreeNode leaf = new PNode(elements);
+        // add leaf to the tree root
         leaf.parent = treeRoot;
-            
     }
-    
+
     /**
-     * TODO: Better Javadoc
-     * Label every positive vertex in the MPQ Tree
-     * 
-     * @param predecessors the predecessors which are used to label the vertices in the tree
-     * @return the labeling of all positive labeled vertices
+     * Get positive labels of the modified PQ-tree
+     *
+     * @return the map of the tree node and the positive label
      */
-    private Map<MPQNode,Integer> labelTree(Set<V> predecessors)
-    {
-        // TODO Auto-generated method stub
+    private Map<MPQTreeNode, Integer> getPositiveLabels(Set<V> predecessors) {
+        labelTree(predecessors);
         return null;
     }
-    
+
+    /**
+     * Label every node N and every section S of a Q-node
+     * according to the relation between the vertex u and the vertices of N or S
+     * <p>
+     * the label is:
+     * - 2, if u is adjacent to all vertices from N or S
+     * - 1, if u is adjacent to some vertices from N or S
+     * - 0, if u is adjacent to no vertices from N or S
+     *
+     * @param predecessors the predecessors which are used to label the vertices in the tree
+     */
+    private void labelTree(Set<V> predecessors) {
+
+    }
+
+    /**
+     * Get the node in the MPQ tree associated with the given vertex in the graph from the map
+     *
+     * @param vertex the vertex in the graph to be tested
+     * @return the associated node in the MPQ tree
+     */
+    private MPQTreeNode getAssociatedTreeNode(V vertex) {
+        Set<MPQTreeNode> associatedPositions = vertexNodeMap.get(vertex);
+
+        // if the associated position set is empty, then the associated tree node is not found
+        if (associatedPositions.isEmpty()) {
+            return null;
+        }
+
+        // if the associated position set contains only one element, then the associated tree node is unique
+        if (associatedPositions.size() == 1) {
+            return associatedPositions.iterator().next();
+        }
+
+        // if the associated position set contains more than one element
```

Comment:
What about QSections? There you can (and most likely will) have two elements here.

Reply:
done by implementing the check for Q section nodes.